README.md
Lightweight implementation of JSON Web Tokens (JWT).

[![Build Status](https://travis-ci.org/corsac-dart/jwt.svg?branch=master)](https://travis-ci.org/corsac-dart/jwt)
[![codecov](https://codecov.io/gh/corsac-dart/jwt/branch/master/graph/badge.svg)](https://codecov.io/gh/corsac-dart/jwt)


## Usage

```dart
import 'package:corsac_jwt/corsac_jwt.dart';

void main() {
  var builder = new JWTBuilder();
  var token = builder
    ..issuer = 'https://api.foobar.com'
    ..expiresAt = new DateTime.now().add(new Duration(minutes: 3))
    ..setClaim('data', {'userId': 836})
    ..getToken(); // returns token without signature

  var signer = new JWTHmacSha256Signer('sharedSecret');
  var signedToken = builder.getSignedToken(signer);
  print(signedToken); // prints encoded JWT
  var stringToken = signedToken.toString();

  var decodedToken = new JWT.parse(stringToken);
  // Verify signature:
  print(decodedToken.verify(signer)); // true

  // Validate claims:
  var validator = new JWTValidator() // uses DateTime.now() by default
    ..issuer = 'https://api.foobar.com'; // set claims you wish to validate
  Set<String> errors = validator.validate(decodedToken);
  print(errors); // (empty list)
}
```

Supported signers:

* HS256 (`JwtHmacSha256Signer`).
* RS256 (`JwtRsaSha256Signer`)

Refer to documentation for more details.

## License

BSD-2


CHANGELOG.md
## 0.4.0

- Allow setting custom headers. See `JWTBuilder.setHeader` for details.

## 0.3.0

- Upgraded dependencies.

## 0.2.2

- Upgraded dependencies which fixes an issue with validating Firebase id tokens.

## 0.2.1

- Added getter to access headers map (read-only).

## 0.2.0

- Added back RS256 signer (#11)

## 0.1.2

- Exposed complete `claims` Map (read-only).

## 0.1.1

- Relaxed SDK constraint to allow 2.0.0-dev versions.

## 0.1.0

- Initial release.


test/all.dart
import 'jwt_test.dart' as jwt;
import 'rs256_test.dart' as rs256;

void main() {
  jwt.main();
  rs256.main();
}


test/jwt_test.dart
library corsac_jwt.test;

import 'package:corsac_jwt/corsac_jwt.dart';
import 'package:test/test.dart';

int _secondsSinceEpoch(DateTime dateTime) {
  return (dateTime.millisecondsSinceEpoch / 1000).floor();
}

void main() {
  group('$JWT', () {
    DateTime now;
    JWTBuilder builder;

    setUp(() {
      now = DateTime.now();
      builder = JWTBuilder();
      builder
        ..issuer = 'https://mycompany.com'
        ..audience = 'people'
        ..issuedAt = now
        ..expiresAt = now.add(Duration(seconds: 10))
        ..notBefore = now.add(Duration(seconds: 5))
        ..id = 'identifier'
        ..subject = 'subj';
    });

    test('JWTError toString', () {
      final error = JWTError('failed');
      expect(error.toString(), 'JWTError: failed');
    });

    test('JWTBuilder can build unsigned token', () {
      var token = builder.getToken();
      expect(token, const TypeMatcher<JWT>());
      expect(token.issuer, equals('https://mycompany.com'));
    });

    test('JWTBuilder can build signed token', () {
      var signer = JWTHmacSha256Signer('secret1');
      var token = builder.getSignedToken(signer);

      expect(token, const TypeMatcher<JWT>());
      expect(token.issuer, equals('https://mycompany.com'));
      expect(token.verify(signer), isTrue);
      expect(token.verify(JWTHmacSha256Signer('invalid')), isFalse);
    });

    test('it parses string token', () {
      var signer = JWTHmacSha256Signer('secret1');
      var stringToken =
          'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL215Y29tcGFueS5jb20ifQ.R7OVbiAKtvSkE-qF0fCkZP_m2JGrHobbRayHhEsKuKU';
      var token = JWT.parse(stringToken);

      expect(token, const TypeMatcher<JWT>());
      expect(token.issuer, equals('https://mycompany.com'));
      expect(token.verify(signer), isTrue);
      expect(token.verify(JWTHmacSha256Signer('invalid')), isFalse);
    });

    test('it parses another token', () {
      var signer = JWTHmacSha256Signer('secret');
      var stringToken =
          'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL215Zm9vYmFyLmNvbSIsImlhdCI6MTQ1NTIzMjI2NywiZXhwIjoxNDU1MjM0MDY3LCJuYmYiOjE0NTUyMzIyMzcsImJvZHkiOnsiYmxhaCI6ImJvb2EifX0.PXbDbE7YapU-6WvRqbdQ2OC1N2DScadvuQUqTHXopNc';

      var token = JWT.parse(stringToken);

      expect(token, const TypeMatcher<JWT>());
      expect(token.issuer, equals('https://myfoobar.com'));
      expect(token.verify(signer), isTrue);
      expect(token.verify(JWTHmacSha256Signer('invalid')), isFalse);
      expect(token.toString(), stringToken);
    });

    test('it throws JWTError if token is invalid', () {
      var badToken1 =
          'invalid.eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwczovL215Y29tcGFueS5jb20ifQ.R7OVbiAKtvSkE-qF0fCkZP_m2JGrHobbRayHhEsKuKU';
      var badToken2 = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.invalid';
      expect(
          () => JWT.parse(badToken1), throwsA(const TypeMatcher<JWTError>()));
      expect(
          () => JWT.parse(badToken2), throwsA(const TypeMatcher<JWTError>()));
    });

    test('it supports all standard claims', () {
      var token = builder.getToken();
      expect(token, const TypeMatcher<JWT>());
      expect(token.issuer, equals('https://mycompany.com'));
      expect(token.audience, equals('people'));
      expect(token.issuedAt, equals(_secondsSinceEpoch(now)));
      expect(token.expiresAt, equals(_secondsSinceEpoch(now) + 10));
      expect(token.notBefore, equals(_secondsSinceEpoch(now) + 5));
      expect(token.id, equals('identifier'));
      expect(token.subject, equals('subj'));
      expect(token.algorithm, equals('none'));
    });

    test('it prevents setting standard claims using setClaim', () {
      expect(() => builder.setClaim('iss', 'bad'), throwsArgumentError);
    });

    test('it supports custom (private) claims', () {
      builder
        ..issuer = 'https://foobar.com'
        ..setClaim('pld', 'payload')
        ..setClaim('map', {'key': 'value'});
      var token = builder.getToken();
      expect(token.issuer, equals('https://foobar.com'));
      expect(token.getClaim('pld'), equals('payload'));
      expect(token.getClaim('map'), equals({'key': 'value'}));

      var stringToken = token.toString();
      var parsedToken = JWT.parse(stringToken);

      expect(parsedToken.issuer, equals('https://foobar.com'));
      expect(parsedToken.getClaim('pld'), equals('payload'));
      expect(parsedToken.getClaim('map'), equals({'key': 'value'}));

      var claims = parsedToken.claims;

      expect(claims['pld'], equals('payload'));
      expect(claims['map'], equals({'key': 'value'}));

      claims['pld'] = 'good times!';
      expect(claims['pld'], equals('good times!'));
      expect(parsedToken.getClaim('pld'), equals('payload'));
    });

    test('it supports custom headers', () {
      builder
        ..issuer = 'https://foobar.com'
        ..setHeader('x5t', 'payload');
      var token = builder.getToken();
      expect(token.issuer, equals('https://foobar.com'));
      expect(token.headers['x5t'], equals('payload'));

      var stringToken = token.toString();
      var parsedToken = JWT.parse(stringToken);

      expect(parsedToken.issuer, equals('https://foobar.com'));
      expect(parsedToken.headers['x5t'], equals('payload'));
    });

    test('it throws error for updating reserved headers', () {
      expect(() {
        builder.setHeader('typ', 'error');
      }, throwsArgumentError);
    });

    test('validator uses current time by default', () {
      final validator = JWTValidator();
      expect(validator.currentTime, isNotNull);
    });

    test('iss claim is validated', () {
      var token = builder.getToken();
      var time = DateTime.now().add(Duration(seconds: 6));
      var validator = JWTValidator(currentTime: time)..issuer = 'wrong';
      var errors = validator.validate(token);
      expect(errors, isNotEmpty);
      expect(errors, contains('The token issuer is invalid.'));

      validator.issuer = 'https://mycompany.com';
      errors = validator.validate(token);
      expect(errors, isEmpty);
    });

    test('exp claim is validated', () {
      var token = builder.getToken();
      var validator =
          JWTValidator(currentTime: DateTime.now().add(Duration(seconds: 20)));
      var errors = validator.validate(token);
      expect(errors, isNotEmpty);
      expect(errors, contains('The token has expired.'));

      validator =
          JWTValidator(currentTime: DateTime.now().add(Duration(seconds: 5)));
      errors = validator.validate(token);
      expect(errors, isEmpty);
    });

    test('iat claim is validated', () {
      var token = builder.getToken();

      var validator = JWTValidator(
          currentTime: DateTime.now().subtract(Duration(seconds: 1)));
      var errors = validator.validate(token);
      expect(errors, isNotEmpty);
      expect(errors, contains('The token issuedAt time is in future.'));

      var time = DateTime.now().add(Duration(seconds: 6));
      validator = JWTValidator(currentTime: time);
      errors = validator.validate(token);
      expect(errors, isEmpty);
    });

    test('nbf claim is validated', () {
      var token = builder.getToken();
      var validator =
          JWTValidator(currentTime: DateTime.now().add(Duration(seconds: 1)));
      var errors = validator.validate(token);
      expect(errors, isNotEmpty);
      expect(errors,
          contains('The token can not be accepted due to notBefore policy.'));

      var time = DateTime.now().add(Duration(seconds: 6));
      validator = JWTValidator(currentTime: time);
      errors = validator.validate(token);
      expect(errors, isEmpty);
    });

    test('aud claim is validated', () {
      var token = builder.getToken();
      var time = DateTime.now().add(Duration(seconds: 6));
      var validator = JWTValidator(currentTime: time)..audience = 'wrong';
      var errors = validator.validate(token);
      expect(errors, isNotEmpty);
      expect(errors, contains('The token audience is invalid.'));

      validator.audience = 'people';
      errors = validator.validate(token);
      expect(errors, isEmpty);
    });

    test('sub claim is validated', () {
      var token = builder.getToken();
      var time = DateTime.now().add(Duration(seconds: 6));
      var validator = JWTValidator(currentTime: time)..subject = 'wrong';
      var errors = validator.validate(token);
      expect(errors, isNotEmpty);
      expect(errors, contains('The token subject is invalid.'));

      validator.subject = 'subj';
      errors = validator.validate(token);
      expect(errors, isEmpty);
    });

    test('jti claim is validated', () {
      var token = builder.getToken();
      var time = DateTime.now().add(Duration(seconds: 6));
      var validator = JWTValidator(currentTime: time)..id = 'wrong';
      var errors = validator.validate(token);
      expect(errors, isNotEmpty);
      expect(errors, contains('The token unique identifier is invalid.'));

      validator.id = 'identifier';
      errors = validator.validate(token);
      expect(errors, isEmpty);
    });

    test('signature is validated', () {
      var signer = JWTHmacSha256Signer('secret');

      var token = builder.getSignedToken(signer);
      var time = DateTime.now().add(Duration(seconds: 6));
      var validator = JWTValidator(currentTime: time);
      var errors =
          validator.validate(token, signer: JWTHmacSha256Signer('invalid'));
      expect(errors, isNotEmpty);
      expect(errors, contains('The token signature is invalid.'));

      errors = validator.validate(token, signer: signer);
      expect(errors, isEmpty);
    });

    test('it provides read-only access to headers', () {
      builder..issuer = 'https://foobar.com';
      var token = builder.getToken();

      var headers = token.headers;
      expect(headers['typ'], 'JWT');
      expect(headers['alg'], 'none');
      expect(() {
        headers['kid'] = 'boom';
      }, throwsUnsupportedError);
    });
  });
}


test/rs256_test.dart
import 'dart:io';

import 'package:corsac_jwt/corsac_jwt.dart';
import 'package:test/test.dart';

void main() {
  group('RS256: ', () {
    JWTRsaSha256Signer signer;
    setUp(() {
      var priv = File('test/resources/private.pem').readAsStringSync();
      var pub = File('test/resources/public.pem').readAsStringSync();
      signer = JWTRsaSha256Signer(privateKey: priv, publicKey: pub);
    });

    test('it can sign and verify JWT with RS256', () {
      var builder = JWTBuilder();
      builder
        ..issuer = 'abc.com'
        ..expiresAt = DateTime.now().add(Duration(minutes: 3));
      var token = builder.getSignedToken(signer);

      expect(token.algorithm, 'RS256');
      expect(token.verify(signer), isTrue);
    });

    test('it handles exceptions on verification', () {
      var jwtString = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiJ9.'
          'eyJpZCI6ImY0MTJiMDViZTRhYTIwZmJlNmEwMDUyZjA5YjdlMzhjOTlmYjdhNjEiLCJqdGkiOiJmNDEyYjA1YmU0YWEyMGZiZTZhMDA1MmYwOWI3ZTM4Yzk5ZmI3YTYxIiwiaXNzIjoiaHR0cDovL2FwaS5mb29iYXIuY29tIn0.'
          'B2vu-KIEFvY3T_EPAYFw48OS7Q7kKVbbOSMIhptyIHZximJ6hkFCBTr2Czz5ArbEYJfAL8_3ZtV3Il6YxE5XQF5hVFNet-Ypt-RzRXPtKMAqt_iiu4C4qg7qes9penNHgu2hvZbQ2FpSPGKrt_ozNehy52YysAKmzKj2ZSelru81ap80pgkYC6Eql8DGIqgz6OVHj_9NRQHqJ2OHDi_nLjYSQW6BtKSA-nmaySr_wn2rMe2xSaf2iA3mPiheCN6yL8yvwcGziNX3wtyahuL1vxg_wJ-sD-py9X7bLu9OmoWds76gxAQQh0Wi694FXQ5p5e4ub0BDlJ9Pv2vr2uPzKL6OQpY4wYBYNhe4UF2QxfFjwnYWITo6O6_tiQtH7Q6WNqF27OfriGYNQbiOgD0icpRrL9w3JI907G4bO3bm0mCIimBbLgr0B_pM4Pr5wcbhXC71yZ0j3ODlfXJ9qnO9G5aAGC9wwFsGT0jgv0ydReDwMgasMN4lYl_iUkxckhHR6ys3wg8FG6SG818CvG-jOkrJMIjNXD9nZMVXPH-tqP-8_60SN8G5vPVdB0nxwL6FprWIc6jC-eXPsATN4E4YJnu5Wnsd4VEPKZVX4Q1AFVOO6dgDxZ7jGStHx50Q1zh1GuNIeEmSnWsVsgtkhTymyZNQSvoIiZnq-wcNtB-YFvY';
      var token = JWT.parse(jwtString);
      expect(token.verify(signer), isFalse);
    });
  });
}


lib/rs256.dart
/// Library implementing JWT RS256 signer.
library corsac_jwt.rs256;

import 'dart:typed_data';

import 'package:bignum/bignum.dart';
import 'package:pointycastle/pointycastle.dart';
import 'package:rsa_pkcs/rsa_pkcs.dart' as rsa;
import 'package:logging/logging.dart';
import 'corsac_jwt.dart';

Logger _logger = new Logger('JWTRsaSha256Signer');

class JWTRsaSha256Signer implements JWTSigner {
  final rsa.RSAPrivateKey _privateKey;
  final rsa.RSAPublicKey _publicKey;

  JWTRsaSha256Signer._(this._privateKey, this._publicKey);

  /// Creates new signer.
  ///
  /// [privateKey] is only required when signing new tokens and otherwise can
  /// be left as `null`. Similarly [publicKey] is only used to verify existing
  /// signatures.
  ///
  /// Both `privateKey` and `publicKey` are expected to be strings in PEM
  /// format.
  factory JWTRsaSha256Signer(
      {String privateKey, String publicKey, String password}) {
    rsa.RSAPKCSParser parser = new rsa.RSAPKCSParser();

    rsa.RSAPrivateKey priv;
    rsa.RSAPublicKey pub;
    if (privateKey is String) {
      rsa.RSAKeyPair pair = parser.parsePEM(privateKey, password: password);
      if (pair.private is! rsa.RSAPrivateKey)
        throw new JWTError('Invalid private RSA key.');
      priv = pair.private;
    }

    if (publicKey is String) {
      rsa.RSAKeyPair pair = parser.parsePEM(publicKey, password: password);
      if (pair.public is! rsa.RSAPublicKey)
        throw new JWTError('Invalid public RSA key.');
      pub = pair.public;
    }
    return new JWTRsaSha256Signer._(priv, pub);
  }

  @override
  String get algorithm => 'RS256';

  @override
  List<int> sign(List<int> body) {
    if (_privateKey == null) {
      throw new StateError(
          'RS256 signer requires private key to create signatures.');
    }
    var s = new Signer('SHA-256/RSA');
    var key = new RSAPrivateKey(_privateKey.modulus,
        _privateKey.privateExponent, _privateKey.prime1, _privateKey.prime2);
    var param = new ParametersWithRandom(
        new PrivateKeyParameter<RSAPrivateKey>(key),
        new SecureRandom("AES/CTR/PRNG"));

    s.init(true, param);
    RSASignature signature = s.generateSignature(new Uint8List.fromList(body));

    return signature.bytes.toList(growable: false);
  }

  @override
  bool verify(List<int> body, List<int> signature) {
    if (_publicKey == null) {
      throw new StateError(
          'RS256 signer requires public key to verify signatures.');
    }

    try {
      var s = new Signer('SHA-256/RSA');
      var key = new RSAPublicKey(
          _publicKey.modulus, new BigInteger(_publicKey.publicExponent));
      var param = new ParametersWithRandom(
          new PublicKeyParameter<RSAPublicKey>(key),
          new SecureRandom("AES/CTR/PRNG"));

      s.init(false, param);
      var rsaSignature = new RSASignature(new Uint8List.fromList(signature));
      return s.verifySignature(new Uint8List.fromList(body), rsaSignature);
    } catch (e) {
      _logger.warning(
          'RS256 token verification failed with following error: ${e}.', e);
      return false;
    }
  }
}


lib/corsac_jwt.dart
library corsac_jwt;

import 'dart:convert';

/// Lightweight JSON Web Token (JWT) implementation.
///
/// ## Usage
///
///     void main() {
///       var builder = new JWTBuilder();
///       var token = builder
///         ..issuer = 'https://api.foobar.com'
///         ..expiresAt = new DateTime.now().add(new Duration(minutes: 3))
///         ..setClaim('data', {'userId': 836})
///         ..getToken(); // returns token without signature
///
///       var signer = new JWTHmacSha256Signer();
///       var signedToken = builder.getSignedToken(signer, 'sharedSecret');
///       print(signedToken); // prints encoded JWT
///       var stringToken = signedToken.toString();
///
///       var decodedToken = new JWT.parse(stringToken);
///       // Verify signature:
///       print(decodedToken.verify(signer, 'sharedSecret')); // true
///
///       // Validate claims:
///       var validator = new JWTValidator() // uses DateTime.now() by default
///         ..issuer = 'https://api.foobar.com'; // set claims you wish to validate
///       Set<String> errors = validator.validate(decodedToken);
///       print(errors); // (empty list)
///     }
///

import 'dart:typed_data';

import 'package:crypto/crypto.dart';
import 'package:logging/logging.dart';
import 'package:pointycastle/pointycastle.dart';
import 'package:rsa_pkcs/rsa_pkcs.dart' as rsa;

Logger _logger = Logger('JWTRsaSha256Signer');

class JWTRsaSha256Signer implements JWTSigner {
  final rsa.RSAPrivateKey _privateKey;
  final rsa.RSAPublicKey _publicKey;

  JWTRsaSha256Signer._(this._privateKey, this._publicKey);

  /// Creates new signer.
  ///
  /// [privateKey] is only required when signing new tokens and otherwise can
  /// be left as `null`. Similarly [publicKey] is only used to verify existing
  /// signatures.
  ///
  /// Both `privateKey` and `publicKey` are expected to be strings in PEM
  /// format.
  factory JWTRsaSha256Signer(
      {String privateKey, String publicKey, String password}) {
    final parser = rsa.RSAPKCSParser();

    rsa.RSAPrivateKey priv;
    rsa.RSAPublicKey pub;
    if (privateKey is String) {
      final pair = parser.parsePEM(privateKey, password: password);
      if (pair.private is! rsa.RSAPrivateKey) {
        throw JWTError('Invalid private RSA key.');
      }
      priv = pair.private;
    }

    if (publicKey is String) {
      final pair = parser.parsePEM(publicKey, password: password);
      if (pair.public is! rsa.RSAPublicKey) {
        throw JWTError('Invalid public RSA key.');
      }
      pub = pair.public;
    }
    return JWTRsaSha256Signer._(priv, pub);
  }

  @override
  String get algorithm => 'RS256';

  @override
  List<int> sign(List<int> body) {
    if (_privateKey == null) {
      throw StateError(
          'RS256 signer requires private key to create signatures.');
    }
    var s = Signer('SHA-256/RSA');
    var key = RSAPrivateKey(_privateKey.modulus, _privateKey.privateExponent,
        _privateKey.prime1, _privateKey.prime2);
    var param = ParametersWithRandom(
      PrivateKeyParameter<RSAPrivateKey>(key),
      SecureRandom('AES/CTR/PRNG'),
    );

    s.init(true, param);
    RSASignature signature = s.generateSignature(Uint8List.fromList(body));

    return signature.bytes.toList(growable: false);
  }

  @override
  bool verify(List<int> body, List<int> signature) {
    if (_publicKey == null) {
      throw StateError(
          'RS256 signer requires public key to verify signatures.');
    }

    try {
      var s = Signer('SHA-256/RSA');
      var key = RSAPublicKey(
          _publicKey.modulus, BigInt.from(_publicKey.publicExponent));
      var param = ParametersWithRandom(
        PublicKeyParameter<RSAPublicKey>(key),
        SecureRandom('AES/CTR/PRNG'),
      );

      s.init(false, param);
      var rsaSignature = RSASignature(Uint8List.fromList(signature));
      return s.verifySignature(Uint8List.fromList(body), rsaSignature);
    } catch (e) {
      _logger.warning(
          'RS256 token verification failed with following error: ${e}.', e);
      return false;
    }
  }
}

final _jsonToBase64Url = json.fuse(utf8.fuse(base64Url));

int _secondsSinceEpoch(DateTime dateTime) {
  return (dateTime.millisecondsSinceEpoch / 1000).floor();
}

String _base64Padded(String value) {
  var mod = value.length % 4;
  if (mod == 0) {
    return value;
  } else if (mod == 3) {
    return value.padRight(value.length + 1, '=');
  } else if (mod == 2) {
    return value.padRight(value.length + 2, '=');
  } else {
    return value; // let it fail when decoding
  }
}

String _base64Unpadded(String value) {
  if (value.endsWith('==')) return value.substring(0, value.length - 2);
  if (value.endsWith('=')) return value.substring(0, value.length - 1);
  return value;
}

/// Error thrown by `JWT` when parsing tokens from string.
class JWTError implements Exception {
  final String message;

  JWTError(this.message);

  @override
  String toString() => 'JWTError: $message';
}

/// JSON Web Token.
class JWT {
  /// List of standard (reserved) claims.
  static const reservedClaims = [
    'iss',
    'aud',
    'iat',
    'exp',
    'nbf',
    'sub',
    'jti'
  ];

  /// List of reserved headers.
  static const reservedHeaders = ['typ', 'alg'];

  /// Allows access to the full headers map.
  ///
  /// Returned map is read-only.
  Map<String, String> get headers => Map.unmodifiable(_headers);
  Map<String, String> _headers;

  /// Allows access to the full claims map.
  ///
  /// Returns modifiable copy of internal map object.
  Map<String, dynamic> get claims => Map.from(_claims);
  Map<String, dynamic> _claims;

  /// Contains original Base64 encoded token header.
  final String encodedHeader;

  /// Contains original Base64 encoded token payload (claims).
  final String encodedPayload;

  /// Contains original Base64 encoded token signature, or `null`
  /// if token is unsigned.
  final String signature;

  JWT._(this.encodedHeader, this.encodedPayload, this.signature) {
    try {
      /// Dart's built-in BASE64URL codec needs padding (SDK 1.17).
      _headers = Map<String, String>.from(
          _jsonToBase64Url.decode(_base64Padded(encodedHeader)));
      _claims = Map<String, dynamic>.from(
          _jsonToBase64Url.decode(_base64Padded(encodedPayload)));
    } catch (e) {
      throw JWTError('Could not decode token string. Error: ${e}.');
    }
  }

  /// Parses [token] string and creates new instance of [JWT].
  /// Throws [JWTError] if parsing fails.
  factory JWT.parse(String token) {
    var parts = token.split('.');
    if (parts.length == 2) {
      return JWT._(parts.first, parts.last, null);
    } else if (parts.length == 3) {
      return JWT._(parts[0], parts[1], parts[2]);
    } else {
      throw JWTError('Invalid token string format for JWT.');
    }
  }

  /// Algorithm used to sign this token. The value `none` means this token
  /// is not signed.
  ///
  /// One should not rely on this value to determine the algorithm used to sign
  /// this token.
  String get algorithm => _headers['alg'];

  /// The issuer of this token (value of standard `iss` claim).
  String get issuer => _claims['iss'];

  /// The audience of this token (value of standard `aud` claim).
  String get audience => _claims['aud'];

  /// The time this token was issued (value of standard `iat` claim).
  int get issuedAt => _claims['iat'];

  /// The expiration time of this token (value of standard `exp` claim).
  int get expiresAt => _claims['exp'];

  /// The time before which this token must not be accepted (value of standard
  /// `nbf` claim).
  int get notBefore => _claims['nbf'];

  /// Identifies the principal that is the subject of this token (value of
  /// standard `sub` claim).
  String get subject => _claims['sub'];

  /// Unique identifier of this token (value of standard `jti` claim).
  String get id => _claims['jti'];

  @override
  String toString() {
    var buffer = StringBuffer();
    buffer.writeAll([encodedHeader, '.', encodedPayload]);
    if (signature is String) {
      buffer.writeAll(['.', signature]);
    }
    return buffer.toString();
  }

  /// Verifies this token's signature using [signer].
  ///
  /// Returns `true` if signature is valid and `false` otherwise.
  bool verify(JWTSigner signer) {
    var body = utf8.encode(encodedHeader + '.' + encodedPayload);
    var sign = base64Url.decode(_base64Padded(signature));
    return signer.verify(body, sign);
  }

  /// Returns value associated with claim specified by [key].
  dynamic getClaim(String key) => _claims[key];
}

/// Builder for JSON Web Tokens.
class JWTBuilder {
  final Map<String, dynamic> _claims = {};
  final Map<String, dynamic> _headers = {'typ': 'JWT', 'alg': 'none'};

  /// Token issuer (standard `iss` claim).
  set issuer(String issuer) {
    _claims['iss'] = issuer;
  }

  /// Token audience (standard `aud` claim).
  set audience(String audience) {
    _claims['aud'] = audience;
  }

  /// Token issued at timestamp in seconds (standard `iat` claim).
  set issuedAt(DateTime issuedAt) {
    _claims['iat'] = _secondsSinceEpoch(issuedAt);
  }

  /// Token expires timestamp in seconds (standard `exp` claim).
  set expiresAt(DateTime expiresAt) {
    _claims['exp'] = _secondsSinceEpoch(expiresAt);
  }

  /// Sets value for standard `nbf` claim.
  set notBefore(DateTime notBefore) {
    _claims['nbf'] = _secondsSinceEpoch(notBefore);
  }

  /// Sets standard `sub` claim value.
  set subject(String subject) {
    _claims['sub'] = subject;
  }

  set id(String id) {
    _claims['jti'] = id;
  }

  /// Sets value of private (custom) claim.
  ///
  /// This method cannot be used to
  /// set values of standard (reserved) claims.
  void setClaim(String name, value) {
    if (JWT.reservedClaims.contains(name.toLowerCase())) {
      throw ArgumentError.value(
          name, 'name', 'Only custom claims can be set with setClaim.');
    }
    _claims[name] = value;
  }

  /// Sets value of a private (custom) header.
  ///
  /// This method cannot be used to update standard (reserved) headers.
  void setHeader(String name, value) {
    if (JWT.reservedHeaders.contains(name.toLowerCase())) {
      throw ArgumentError.value(
          name, 'name', 'Only custom headers can be set with setHeader.');
    }
    _headers[name] = value;
  }

  /// Builds and returns JWT. The token will not be signed.
  ///
  /// To create signed token use [getSignedToken] instead.
  JWT getToken() {
    final encodedHeader = _base64Unpadded(_jsonToBase64Url.encode(_headers));
    final encodedPayload = _base64Unpadded(_jsonToBase64Url.encode(_claims));
    return JWT._(encodedHeader, encodedPayload, null);
  }

  /// Builds and returns signed JWT.
  ///
  /// The token is signed with provided [signer].
  ///
  /// To create unsigned token use [getToken].
  JWT getSignedToken(JWTSigner signer) {
    _headers['alg'] = signer.algorithm;
    final encodedHeader = _base64Unpadded(_jsonToBase64Url.encode(_headers));
    final encodedPayload = _base64Unpadded(_jsonToBase64Url.encode(_claims));
    var body = encodedHeader + '.' + encodedPayload;
    var signature =
        _base64Unpadded(base64Url.encode(signer.sign(utf8.encode(body))));
    return JWT._(encodedHeader, encodedPayload, signature);
  }
}

/// Signer interface for JWT.
abstract class JWTSigner {
  String get algorithm;
  List<int> sign(List<int> body);
  bool verify(List<int> body, List<int> signature);
}

/// Signer implementing HMAC encryption using SHA256 hashing.
class JWTHmacSha256Signer implements JWTSigner {
  final List<int> secret;

  JWTHmacSha256Signer(String secret) : secret = utf8.encode(secret);

  @override
  String get algorithm => 'HS256';

  @override
  List<int> sign(List<int> body) {
    final hmac = Hmac(sha256, secret);
    return hmac.convert(body).bytes;
  }

  @override
  bool verify(List<int> body, List<int> signature) {
    var actual = sign(body);
    if (actual.length == signature.length) {
      // constant-time comparison
      var isEqual = true;
      for (var i = 0; i < actual.length; i++) {
        if (actual[i] != signature[i]) isEqual = false;
      }
      return isEqual;
    } else {
      return false;
    }
  }
}

/// Validator for JSON Web Tokens.
///
/// One must configure validator and provide values for claims that should be
/// validated, except for `iat`, `exp` and `nbf` claims - these are always
/// validated based on the value of [currentTime].
class JWTValidator {
  /// Current time used to validate token's `iat`, `exp` and `nbf` claims.
  final DateTime currentTime;
  String issuer;
  String audience;
  String subject;
  String id;

  /// Creates new validator. One can supply custom value for [currentTime]
  /// parameter, if not the `DateTime.now()` value is used by default.
  JWTValidator({DateTime currentTime})
      : currentTime = currentTime ?? DateTime.now();

  /// Validates provided [token] and returns a list of validation errors.
  /// Empty list indicates there were no validation errors.
  ///
  /// If [signer] parameter is provided then token signature
  /// will also be verified. Otherwise signature must be verified manually using
  /// [JWT.verify] method.
  Set<String> validate(JWT token, {JWTSigner signer}) {
    var errors = <String>{};

    var currentTimestamp = _secondsSinceEpoch(currentTime);
    if (token.expiresAt is int && currentTimestamp >= token.expiresAt) {
      errors.add('The token has expired.');
    }

    if (token.issuedAt is int && currentTimestamp < token.issuedAt) {
      errors.add('The token issuedAt time is in future.');
    }

    if (token.notBefore is int && currentTimestamp < token.notBefore) {
      errors.add('The token can not be accepted due to notBefore policy.');
    }

    if (issuer is String && issuer != token.issuer) {
      errors.add('The token issuer is invalid.');
    }

    if (audience is String && audience != token.audience) {
      errors.add('The token audience is invalid.');
    }

    if (subject is String && subject != token.subject) {
      errors.add('The token subject is invalid.');
    }

    if (id is String && id != token.id) {
      errors.add('The token unique identifier is invalid.');
    }

    if (signer is JWTSigner && !token.verify(signer)) {
      errors.add('The token signature is invalid.');
    }

    return errors;
  }
}
